[OutOfMemoryError] Fix OOM at `LogEncrypt.encrypt(...)`

ParaskP7 opened the pull request:
Fixes: #16
Closes: [AINFRA-1217](https://linear.app/a8c/issue/AINFRA-1217)
(Internal) Discussions: p1754667446732699-slack-C030U03RC8Y + p1757585328963649-slack-C030U03RC8Y

---

### Description

This PR resolves a long standing `OutOfMemory` (OOM) error when using this library alongside logs that are of large size.

FYI: Actually, this change resolves a couple more (potential) issues like:
- `ANR`s when the library tries to process large logs (no matter if that results in an OOM or not). See for example: [AINFRA-1182](https://linear.app/a8c/issue/AINFRA-1182)
- `InvalidRequest`s when the library is successful in processing large logs (`10MB+`), but the API respond with `400` just because of the [MAX_FILE_SIZE_IN_BYTES](https://github.a8c.com/Automattic/wpcom/blob/trunk/public.api/rest/wpcom-json-endpoints/class.wpcom-json-api-encrypted-logs-endpoint.php#L12C16-L12C38) restriction on [WPCOM_JSON_API_Encrypted_Logs_Endpoint](https://github.a8c.com/Automattic/wpcom/blob/trunk/public.api/rest/wpcom-json-endpoints/class.wpcom-json-api-encrypted-logs-endpoint.php).

Overall, this change:
1. Introduces a `MAX_LOG_FILE_SIZE` of `5MB`, just because the resulting encrypted text is about `1.4` larger per `MB` of a log file, resulting in a final max of about `7MB` that is sent to the API. We could have used `6MB` as the max log file size, which would result in `8.4MB` of encrypted text sent, but it is better to keep enough buffer just in case, and this way, to safeguard ourselves from any `InvalidRequest` responses from the API.
2. Based on this `MAX_LOG_FILE_SIZE` of `5MB`, if a log's file size exceeds this max, it is truncated to that `5MB` of size, skipping the beginning of the log and only keep the more recent logs.

---

### Testing Steps

Using the [DOAndroid](https://github.com/bloom/DayOne-Android) project, update `automattic-encryptedlogging` to `28-f19c0d92473042e5976c92cd7b88cf641456691a` (version which is based on this PR's [latest](https://github.com/Automattic/EncryptedLogging/pull/28/commits/f19c0d92473042e5976c92cd7b88cf641456691a) commit) and make sure that encrypted logging could work with logs of large size.

> [!NOTE]
> I would first advise you to replicate an OOM using your physical/virtual device of choice. To do so, update `automattic-encryptedlogging` to `28-a16584d1c883c758c29e8e40f8bba0c1320ff567` instead. This commit skips the uploading of the encrypted log to the API, and just handles the encryption part.

> [!TIP]
> FYI: You could use the below [DOAndroid](https://github.com/bloom/DayOne-Android) patch, which add the `Fill The Log` development option into the `Developer` screen of the app and on click adds `10000` log lines (or else `4MB` worth of logs). Adding another `0` to that `numberOfLogs` val will give you `40MB` instead, use what you need:
> 
> <details>
>     <summary>patch</summary>
> 
> ```diff
> Subject: [PATCH] Tmp: Assemble release on ci for pr
> ---
> Index: app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperAction.kt
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> diff --git a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperAction.kt b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperAction.kt
> --- a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperAction.kt	(revision f2622054d961b58bf298d4ab91049be79c19a87b)
> +++ b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperAction.kt	(date 1757609295757)
> @@ -14,6 +14,8 @@
>  
>      data object CrashApp : DeveloperAction()
>  
> +    data object FillLog : DeveloperAction()
> +
>      data object DeleteAllJournals : DeveloperAction()
>  
>      data object ReloadCurrentMasterKeyStorage : DeveloperAction()
> Index: app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperFragment.kt
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> diff --git a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperFragment.kt b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperFragment.kt
> --- a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperFragment.kt	(revision f2622054d961b58bf298d4ab91049be79c19a87b)
> +++ b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperFragment.kt	(date 1757609295757)
> @@ -141,6 +141,20 @@
>                  }
>              }
>              DeveloperAction.CrashApp -> throw Exception("Developer caused a crash!")
> +            DeveloperAction.FillLog -> {
> +                val numberOfLogs = 10000 // 4MB
> +                (0..numberOfLogs).forEachIndexed { index, _ ->
> +                    DOLogger.d(TAG, "Filling log line $index " +
> +                            "[FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST " +
> +                            "FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST " +
> +                            "FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST " +
> +                            "FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST FILL LOG TEST]"
> +                    )
> +                    if (index % (numberOfLogs/5) == 0) {
> +                        Toast.makeText(context, "Filled $index log lines", Toast.LENGTH_SHORT).show()
> +                    }
> +                }
> +            }
>              DeveloperAction.DeleteAllJournals -> deleteAllJournals()
>              DeveloperAction.ReloadCurrentMasterKeyStorage -> viewModel.reloadMasterKeyStorage()
>              DeveloperAction.EnterMasterKey -> enterMasterKey()
> Index: app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperComponentList.kt
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> diff --git a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperComponentList.kt b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperComponentList.kt
> --- a/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperComponentList.kt	(revision f2622054d961b58bf298d4ab91049be79c19a87b)
> +++ b/app/src/main/java/com/dayoneapp/dayone/main/settings/DeveloperComponentList.kt	(date 1757609295757)
> @@ -413,6 +413,10 @@
>          title = { Text(text = stringResource(R.string.crash_app)) },
>          onClick = { onClick(DeveloperAction.CrashApp) },
>      )
> +    SettingItem(
> +        title = { Text(text = stringResource(R.string.fill_log)) },
> +        onClick = { onClick(DeveloperAction.FillLog) },
> +    )
>      if (BuildConfig.DEBUG) {
>          SettingItem(
>              title = { Text(text = stringResource(R.string.purge_welcome_entries)) },
> Index: app/src/main/res/values/strings.xml
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> diff --git a/app/src/main/res/values/strings.xml b/app/src/main/res/values/strings.xml
> --- a/app/src/main/res/values/strings.xml	(revision f2622054d961b58bf298d4ab91049be79c19a87b)
> +++ b/app/src/main/res/values/strings.xml	(date 1757609295758)
> @@ -881,6 +881,7 @@
>      <string name="reset_passive_message_settings">Reset Passive Messages Settings</string>
>      <string name="reset_today_tooltip_message_settings">Reset Today Tooltip Message Settings</string>
>      <string name="crash_app">Crash The App</string>
> +    <string name="fill_log">Fill The Log</string>
>      <string name="enable_premium_dev">Enable Premium Dev</string>
>      <string name="should_loggedin_premium_dev">You Should be Logged In To Enable Premium Dev</string>
>      <string name="purge_welcome_entries">Purge possible welcome entries</string>
> ```
> 
> </details>
> 
> PS: Don't forget to use the `Device Explorer` to check on the actual size of the log right before you cause a crash (using another `Crash The App` development option from the `Developer` screen of the app. File log path: `/data/data/com.dayoneapp.dayone.debug/files/dayonelogs/com.dayoneapp.dayone.debug YYYY-MM-DD.log`

Follow the below generic steps with a log of `2MB`, `10MB`, `50MB` and `200MB` file size:

1. Using `App Inspection` from AS find the `encrypted-logging.db` and its sole `EncryptedLogModel` table. Notice that the table is empty, there are no log to be sent.
2. Make the app crash. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could navigate to `Settings` -> `Developer` and then click `Crash The App`.
3. You might now notice a new db entry within `App Inspection` from AS on that `DETACHED` app (or not, it depends). Now, switch to the `ATTACHED` app and notice the db empty. This is because this db entry/log has been already sent to Sentry already, on app start and after the app crashed. To really notice the entry, trying crashing the app again, this time put the device on offline mode first. Doing that you'll be able to see the entry on the db, even when switching to the newly created `ATTACHED` app.
4. Open [Sentry](https://a8c.sentry.io) and navigate to `Issues`. Find the project you're testing for (for example `day-one-android` or/and `pocket-casts-android`) and check the latest `debug` exceptions. Using [DOAndroid](https://github.com/bloom/DayOne-Android) for example, you could search for `Developer caused a crash!` and check the latest crash. Now navigate at the bottom and check the `uuid`. Verify it is the same `UUID` that you saw on that db entry before.
5. Open [Read Encrypted Logs](https://mc.a8c.com/encrypted-logs.php), copy-paste that `UUID` into `Log UUID`, click `View` and make sure you can read the newly uploaded and now decrypted log.

For a log of specific size (see above) expect to get the following results, all without any OOM, or ANR, or an `InvalidRequest` response from the API, and the app should launch instantly, that is, after the crash, without any delay:

1. `2MB`:
    - ℹ️ Enqueueing sending encrypted logs. Log file: com.dayoneapp.dayone.debug YYYY-MM-DD.log, 2.00mb
    - ℹ️ The size of encrypted text with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is: 2.80mb
    - ℹ️ Successfully uploaded encrypted log with uuid: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
2. `10MB`:
    - ℹ️ Enqueueing sending encrypted logs. Log file: com.dayoneapp.dayone.debug YYYY-MM-DD.log, 10.00mb
    - ⚠️ Log file with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is too big (10.00mb), max allowed log file size is 5.00mb; log file got truncated.
    - ℹ️ The size of encrypted text with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is: 7.00mb
    - ℹ️ Successfully uploaded encrypted log with uuid: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
3. `50MB`:
    - ℹ️ Enqueueing sending encrypted logs. Log file: com.dayoneapp.dayone.debug YYYY-MM-DD.log, 50.00mb
    - ⚠️ Log file with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is too big (50.00mb), max allowed log file size is 5.00mb; log file got truncated.
    - ℹ️ The size of encrypted text with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is: 7.00mb
    - ℹ️ Successfully uploaded encrypted log with uuid: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX
4. `200MB`:
    - ℹ️ Enqueueing sending encrypted logs. Log file: com.dayoneapp.dayone.debug YYYY-MM-DD.log, 200.00mb
    - ⚠️ Log file with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is too big (200.00mb), max allowed log file size is 5.00mb; log file got truncated.
    - ℹ️ The size of encrypted text with uuid XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX is: 7.00mb
    - ℹ️ Successfully uploaded encrypted log with uuid: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX

ParaskP7 committed:
Tmp: Don't upload encrypted log, just handle the encryption part

FYI: This change is for debugging purposes so that to avoid uploading
log files that are too big, and until we figure out how big a log file
should be to cause this OOM problem. Then, to try to fix the problem and
see if 'logEncrypter.encrypt(...)' could now handle such big log files.

ParaskP7 committed:
Fix: Add logs size limit and pre-trip logs before encrypting

FYI: This 5MB max logs size limit is both arbitrary and empirical
because it seems that for (at least) some standard devices, large logs
files of 100MB+ can easily cause OOM. Also, this 5MB max logs size limit
is chosen because the API accepts logs up to 10MB, but we leave enough
headroom for encoding overhead.

PS: It is worth to note that we could apply a different kind of a
solution (better one), like streaming encryption or chunked processing.
But, I don't think it is worth it, nor I think it is worth sending such
large log files to the API.

ParaskP7 committed:
Revert "Tmp: Don't upload encrypted log, just handle the encryption part"

This reverts commit a16584d1c883c758c29e8e40f8bba0c1320ff567.

ParaskP7 committed:
Fix: Add in-memory size limit and truncate file before encrypting

FYI: Even after fixing the logs size problem during encryption, if the
file is too big (200MB+), just because of the 'file.readText()', which
tries to read the entire file into memory as string, another OOM problem
occurs, this time before the encryption logic even runs.

By file size checking and truncating the file to the last 10MB (if too
big), we are avoiding this OOM issue. The current approach:
- Creates a temp file with only the last 10MB of data.
- Reads the temp file content and returns it.
- Cleans up the temp file in the finally block.
- Leaves the original file completely untouched.

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -151,6 +167,25 @@ internal class EncryptedLogStore private constructor(
+            // Calculate how many bytes to skip from the beginning to keep last MAX_IN_MEMORY_SIZE
+            val skipBytes = file.length() - MAX_IN_MEMORY_SIZE
+            file.inputStream().use { input ->
+                input.skip(skipBytes)
```
We don't need to skip the exact number of bytes requested, so all good here, thanks! 👍

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/model/encryptedlogging/LogEncrypter.kt:
```diff
@@ -17,27 +22,40 @@ internal data class EncryptedLoggingKey(val publicKey: Key)
+        val trimmedText = if (text.length > MAX_LOG_SIZE) {
+            Log.w(TAG, "Log with uuid $uuid is too big (${text.length.toMB()}mb), " +
+                    "max allowed log size is ${MAX_LOG_SIZE.toMB()}mb; log got trimmed.")
+            text.takeLast(MAX_LOG_SIZE) // Keep the last N bytes (most recent logs)
```
Hmmm, right, maybe that's why I was getting an `InvalidRequest` response from the server, even when I used `9MB` as the max log size, maybe that's it... 🤔

ParaskP7 replied:
FYI: This whole `text trimming` fix 6990d7c3b2839ac7ecc2e536d557b1c1e869b995 commit got reverted b0079442b9d010e964f71ec9fbb5815c3f4279a7 and we now only use this `file truncation` fix 7ae6f7dc51c3114518fe0d16ce54cfa153af326d commit with a max log file size of 5MB (a999320debf16014022de7df6e1122f587d613a8).

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -151,6 +167,25 @@ internal class EncryptedLogStore private constructor(
     }
 
+    private fun truncateFile(file: File, uuid: String): String {
+        val tempFile = File.createTempFile("truncated_", ".log", file.parentFile)
```
No, I think this is fine, and as long as this temporary file is then deleted, during the `finally` block, this is perfectly fine, no need to worry too much about it. 👍

ParaskP7 committed:
Revert "Fix: Add logs size limit and pre-trip logs before encrypting"

This reverts commit 6990d7c3

ParaskP7 committed:
Refactor: Change max in memory size to max log file size of 5mb

Now that the initial fix (6990d7c3b2839ac7ecc2e536d557b1c1e869b995) got
reverted (b0079442b9d010e964f71ec9fbb5815c3f4279a7) there is no point
referring to a 'MAX_IN_MEMORY_SIZE' constant, but rather a
'MAX_LOG_FILE_SIZE' constant which also adheres to what the API accepts,
which is logs up to 10MB.

FYI: Although the API accepts logs up to 10MB, we use 5MB to leave
enough room for the encryption overhead (and in order to avoid 400
failure with upload turning into invalid requests).

PS: Actually, while testing this solution I also tried working with 9MB,
then 8MB, then 7MB, all resulting in 400 ('InvalidRequest'). Only when I
tried using 6MB the API started accepting the truncated encrypted log.
As such, an in order to give enough buffer, I set the final value to
5MB, just in case 6MB isn't enough for every case.

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -126,17 +129,30 @@ internal class EncryptedLogStore private constructor(
+                text = encryptedLog.file.readText(),
+                uuid = encryptedLog.uuid
+            )
+        } else { // Large files: extract only the last MAX_IN_MEMORY_SIZE
```
Resolved: a999320debf16014022de7df6e1122f587d613a8

ParaskP7 commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -151,6 +167,25 @@ internal class EncryptedLogStore private constructor(
+    private fun truncateFile(file: File, uuid: String): String {
+        val tempFile = File.createTempFile("truncated_", ".log", file.parentFile)
+        try {
+            // Calculate how many bytes to skip from the beginning to keep last MAX_IN_MEMORY_SIZE
```
[SAME](https://github.com/Automattic/EncryptedLogging/pull/28#discussion_r2341440591)

ParaskP7 committed:
Debug: Check on the size of the encrypted text right before upload

harrytmthy commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/store/EncryptedLogStore.kt:
```diff
@@ -126,17 +129,31 @@ internal class EncryptedLogStore private constructor(
-            val encryptedText = logEncrypter.encrypt(text = encryptedLog.file.readText(), uuid = encryptedLog.uuid)
+        val encryptedText = if (encryptedLog.file.length() <= MAX_LOG_FILE_SIZE) {
+            logEncrypter.encrypt(
+                text = encryptedLog.file.readText(),
```
We cap the file at 5 MB but still load it into memory with `readText()`. This inflates size (UTF-16 + base64) and may still cause OOM on low-RAM devices. Could `LogEncrypter.encrypt` support ByteArray or InputStream for streaming/chunked encryption to avoid building a full String?

ParaskP7 replied:
> We cap the file at 5 MB but still load it into memory with readText(). This inflates size (UTF-16 + base64) and may still cause OOM on low-RAM devices.

That's true @harrytmthy and thanks for the note here! 💯 

> Could LogEncrypter.encrypt support ByteArray or InputStream for streaming/chunked encryption to avoid building a full String?

We could, we'll try this first, since the API anyway doesn't accept encrypted log files larger than `10MB` and take it from there. 👍

ParaskP7 commented:
👋 @wzieba just a friendly reminder about this! 🙏

wzieba commented on encryptedlogging/src/main/kotlin/com/automattic/encryptedlogging/utils/Utils.kt:
```diff
@@ -0,0 +1,17 @@
+        "%.2f",
+        this.toFloat() / 1024.0 / 1024.0
+    )
+}
```
nitpick: if `object` is internal, I believe its methods don't have to

```suggestion
internal object Utils {
    fun Int.toMB() = String.format(
        Locale.US,
        "%.2f",
        this.toFloat() / 1024.0 / 1024.0
    )

    fun Long.toMB() = String.format(
        Locale.US,
        "%.2f",
        this.toFloat() / 1024.0 / 1024.0
    )
}
```

ParaskP7 replied:
Just committed the suggestion @wzieba , thanks: ae25826f9152c580be9429a5546dbc065fa41f30

wzieba reviewed (approved):
LGTM, leaving one minor nitpick

ParaskP7 committed:
Refactor: Remove unnecessary internal keyword on utils functions

PR Comment: https://github.com/Automattic/EncryptedLogging/pull/28
#discussion_r2355022277

Co-authored-by: Wojciech Zięba <wojtek.zieba@automattic.com>

ParaskP7 commented:
Thanks for reviewing/testing and leaving your suggestions on this @wzieba , much appreciated! 🙇 ❤️ 🚀
